### PR TITLE
feat: authentication

### PR DESCRIPTION
**Pull request by davelsan**
## Description

This PR is a rewrite of the authentication logic and related UI components. The goal is to provide a central authentication service, the tools to access it, and a new redux store foundation to build upon.

### Authentication

- Add a new store configuration based on `@reduxjs/toolkit` that is more compartmentalized.
- Add a new `auth-slice` store to contain the authentication state.
- Add new `auth` utilities to retrieve and manage credentials from the remote or localStorage.
- Add a custom `useAuth` hook to access credentials and dispatch login/logout actions.
- Implement the new authentication service in `LoginPage`, `PrivateRoute`, and `Topbar` (logout) components.

### Refactor

- Refactor `Topbar` and `LanguageSwitcher` to use TypeScript.
- Export named constants from `config/globals.ts`.
- Change the material-ui spacing scale to use `rem` instead of `px` (easier support for responsive design).
- Rewrite the `LoginPage` component into several sub-components with more encapsulated logic.
- Remove `.gitkeep` files from non-empty folders.

### Fixes
- Fix a material-ui warning related to `findDOMNode` while using `React.Strict`.
- Fix model routes not being matched before the `ModelHome` component.

### Dependencies

- Add `@reduxjs/toolkit` for state management (abstraction on top of our `redux` implementation).

## Notes

The `LoginPage` component now uses a form with two custom input fields. To avoid conflicts with the incoming `AttributesForm`  PR, which also adds custom input fields, I've named these as `LoginField` (custom text field with an icon) and `PasswordField` (a `LoginField` wrapper with adornment functionality). These should be reviewed again in the future to ensure we don't have code duplication.

### CODE REVIEW

File: src/components/input/PasswordField.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  return (
+    <StringField
+      {...props}
+      type={state.showPassword ? 'text' : 'password'}
+      autoComplete="off"
+      InputProps={{
+        endAdornment: (
+          <InputAdornment position="end">
+            <Tooltip title="Show Password">
```
Comment (vsuaste) [inline src/components/input/PasswordField.tsx:38]: Suggestion: tooltip could toggle between "Show Password" and "Hide password"

**Review by vsuaste — COMMENTED** [review]
**Done** in this review:
- Check functionality:  login/logout properly

**NOT DONE** in this review:

- Check standards for typescript (as I'm not familiar with it).
- Check if authentication was done under the standards(or recommended way) for a react SPA.
- Understand every single line of the code.

**TO FIX**

- LoginField.tsx.tsx has double tsx extension.

**SUGGESTIONS**

- Toggle between show password/ hide password as tooltip in the show/not show password icon. (A comment in the correspondent file was left.)
- It would be nice to add the Zendro logo in the login form.

**TO BE DISCUSSED**

- When the token is already expired or deleted, the routes(home, etc) don't redirect to login page.
- Language switch doesn't working yet.
- I get a warning on the browser related to react-router (I think). I attach the screen shot here.
WarningReactRouter

**Comment by davelsan** [comment]
Thanks for the review, @vsuaste. I've made a few changes based on your comment.

## Changes

- Remove duplicated `LoginField.tsx.tsx` extension.
- Add a conditional "Show Password" / "Hide Password" tooltip, depending on the current password display.

## Discussion

> It would be nice to add the Zendro logo in the login form.

Logo support in the login screen makes sense to me. We could add the Zendro logo as an example, to be replaced or removed by the actual data warehouse user.

> When the token is already expired or deleted, the routes(home, etc) don't redirect to login page.

That the user is not logged off when the token is deleted manually in devtools is a design choice. My reasoning for this is that the token is only used in the initial page load and after other reloads. During a continued session (no reload), the token is stored in memory, at the redux store.

Reaching the expiration time does trigger a token cleanup and redirect to login. The `exp` property is currently set by the server at one hour after `iat`. I've tested this with a shorter internal (5 seconds) and it seems to function as intended.

I've made a minor change in cbb9c0f6f575eda6e5b3639d389e4876b85f2e46 to move the expiration date computation outside of the getter. That way we don't compute it every time `isValid` is called.

> Language switch doesn't work yet.

We need a new PR for this. Could be part of the i18n service refactor?
Currently, all translations are in a monolithic `i18n.resources.js`. file. We could use separate resource files for each translation (`en.json`, `es.json`, `de.json`) and start building from there.

> I get a warning on the browser related to react-router (I think). I attach the screen shot here.

This is present in the `develop` branch as well. It is originated in the `NavGroup` component, because the `Link` component of `react-router-dom` requires a `to` property to be set.

I already fixed a generic warning from `@material-ui/core` in db2422fc9807973ce8b0573d3f66f7fd04a4a67d, but I feel the one you mention requires a slightly larger refactor. I'd suggest a different PR or hotfix to handle it.

File: src/components/input/PasswordField.tsx (hunk 1 above)
Comment (davelsan) [reply src/components/input/PasswordField.tsx]: Added in  98d0c42.

**Review by vsuaste — APPROVED** [review]

**Pull request merged**